docs/source/60_building_apps.md

```diff
@@ -1,3 +1,250 @@
+that assembles and delegates to the various CLI actions.
+
+To add a new command:
+
+Extend `Args4jBase` class to specify arguments to the command. Arguments are defined using
+the [args4j library](http://args4j.kohsuke.org/). If reading or writing to Parquet, consider
+including Parquet arguments via `with ParquetArgs`.
+
+```scala
+class MyCommandArgs extends Args4jBase with ParquetArgs {
+  @Argument(required = true, metaVar = "INPUT", usage = "Input to my command", index = 0)
+  var inputPath: String = null
+}
+```
+
+Extend `BDGCommandCompanion` object to specify the command name and description. The `apply`
+method associates `MyCommandArgs` defined above with `MyCommand`.
+
+```scala
+object MyCommand extends BDGCommandCompanion {
+  val commandName = "myCommand"
+  val commandDescription = "My command example."
+
+  def apply(cmdLine: Array[String]) = {
+    new MyCommand(Args4j[MyCommandArgs](cmdLine))
+  }
+}
+```
+
+Extend `BDGSparkCommand` class and implement the `run(SparkContext)` method. The `MyCommandArgs`
+class defined above is provided in the constructor and specifies the generic type for `BDGSparkCommand`.
+The companion object defined above is declared as a field.  For access to the Apache Spark
+[slf4j](http://www.slf4j.org/) Logger via the `log` field, specify `with Logging`.
+
+```scala
+class MyCommand(protected val args: MyCommandArgs) extends BDGSparkCommand[MyCommandArgs] with Logging {
+  val companion = MyCommand
+
+  def run(sc: SparkContext) {
+    log.info("Doing something...")
+    // do something
+  }
+}
+```
+
+Add the new command to the default list of commands in `ADAMMain`.
+
+```scala
+  val defaultCommandGroups =
+    List(
+      CommandGroup(
+        "ADAM ACTIONS",
+        List(
+          MyCommand,
+          CountReadKmers,
+          CountContigKmers, ...
+```
+
+Build ADAM and run the new command via `adam-submit`.
+
+```bash
+$ mvn install
+$ ./bin/adam-submit --help
+Using ADAM_MAIN=org.bdgenomics.adam.cli.ADAMMain
+Using SPARK_SUBMIT=/usr/local/bin/spark-submit
+
+       e         888~-_          e             e    e
+      d8b        888   \        d8b           d8b  d8b
+     /Y88b       888    |      /Y88b         d888bdY88b
+    /  Y88b      888    |     /  Y88b       / Y88Y Y888b
+   /____Y88b     888   /     /____Y88b     /   YY   Y888b
+  /      Y88b    888_-~     /      Y88b   /          Y888b
+
+Usage: adam-submit [<spark-args> --] <adam-args>
+
+Choose one of the following commands:
+
+ADAM ACTIONS
+           myCommand : My command example.
+          countKmers : Counts the k-mers/q-mers from a read dataset.
+    countContigKmers : Counts the k-mers/q-mers from a read dataset.
+...
+
+$ ./bin/adam-submit myCommand input.foo
+```
+
+Then consider making a pull request to include the new command in ADAM!
+
+
+## Extend the ADAM CLI by adding new commands in an external repository {#external-commands}
+
+To extend the ADAM CLI by adding new commands in an external repository,
```

Comment:
This is the pattern that works through dependency injection, right? Would it be useful to link to the dependency injection tool that we use, and how we use it? I'm just thinking that if someone made a subtle error here (not sure what, but a hypothetical error) that caused the command discovery to fail, this might be useful info for debugging. Perhaps we could add this in a subsection, or something like that.